op-node/rollup/derive/l2block_util.go

```diff
@@ -0,0 +1,64 @@
+package derive
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+)
+
+// L2BlockRefSource is a source for the generation of a L2BlockRef. E.g. a
+// *types.Block is a L2BlockRefSource.
+//
+// L2BlockToBlockRef extracts L2BlockRef from a L2BlockRefSource. The first
+// transaction of a source must be a Deposit transaction.
+type L2BlockRefSource interface {
+	Hash() common.Hash
+	ParentHash() common.Hash
+	NumberU64() uint64
+	Time() uint64
+	Transactions() types.Transactions
+}
+
+// PayloadToBlockRef extracts the essential L2BlockRef information from an L2
+// block ref source, falling back to genesis information if necessary.
+func L2BlockToBlockRef(block L2BlockRefSource, genesis *rollup.Genesis) (eth.L2BlockRef, error) {
```

Comment:
I assume this is copied from `PayloadToBlockRef`. Fine for now, but we are starting to get a proliferation of types.

Reply:
It is. I actually also implemented a converter for an `ExecutionPayload` into a `L2BlockRefSource` and then this could be reused, but then discarded that as too fancy.

Reply:
The reason I added this is to avoid a marshalling roundtrip. A block could also first be converted into an exec payload. But I found that inefficient.